sc2/unit.py

```diff
@@ -295,6 +295,121 @@ def movement_speed(self) -> float:
         # TODO: use bot object to calculate if unit is zerg unit and if it is on creep, and if it your own unit: check for movement speed upgrades (e.g. zergling, hydra, ultralisk) or buffs (stimpack, time warp slow, fungal), perhaps write a second property function for this
         return self._type_data._proto.movement_speed
 
+    @property
+    def real_speed(self):
+        return self.calculate_speed()
+
+    def calculate_speed(self, upgrades=None) -> float:
+        """ Calculates the movement speed of the unit including buffs and upgrades.
+        Note: Upgrades only work with own units. Use "upgrades" param to set expected enemy upgrades.
+
+        :param upgrades: """
+        speed = self.movement_speed
+        unit_type = self.type_id
+
+        # ---- Upgrades ----
+        if upgrades is None:
+            upgrades = self._bot_object.state.upgrades
+
+        if upgrades:
+            if unit_type == UnitTypeId.BANSHEE and UpgradeId.BANSHEESPEED in upgrades:
+                speed = 5.25
+            elif unit_type == UNIT_OBSERVER and UpgradeId.OBSERVERGRAVITICBOOSTER in upgrades:
+                speed *= 2
+            elif unit_type == UnitTypeId.WARPPRISM and UpgradeId.GRAVITICDRIVE in upgrades:
+                speed += 0.422
+            elif unit_type == UnitTypeId.ROACH and UpgradeId.GLIALRECONSTITUTION in upgrades:
+                speed = 4.2
+            elif unit_type == UnitTypeId.OVERLORD and UpgradeId.OVERLORDSPEED in upgrades:
+                speed = 1.88
+            elif unit_type == UnitTypeId.OVERSEER and UpgradeId.OVERLORDSPEED in upgrades:
+                speed = 3.375
+            elif unit_type == UnitTypeId.ZERGLING and UpgradeId.ZERGLINGMOVEMENTSPEED in upgrades:
+                speed *= 1.6
+            elif unit_type == UnitTypeId.BANELING and UpgradeId.CENTRIFICALHOOKS in upgrades:
+                speed += 0.63
+            '''
+            elif unit_type == UnitTypeId.LURKERMP and UpgradeId.DIGGINGCLAWS in upgrades:
+                pass  # todo: can't find how much it increases speed
+            '''
```

Comment:
Digging claws only increases the speed how fast a lurker burrows.

Reply:
No, it really increases their movement speed. Description of upgrade confirms it.
Just checked it in game: speed increases from 4.13 to 4.55

Reply:
increase factor = 4.55 / 4.13 = 1.1016949152542372
Then: `speed *= 1.1016949152542372`

Reply:
[after upgrade](https://i.imgur.com/vmnjcJM.jpg)

Reply:
proofs:
[speed before upgrade](https://i.imgur.com/lpDzhC5.jpg)
[speed after upgrade](https://i.imgur.com/vmnjcJM.jpg)
[upgrade description](https://i.imgur.com/CeGLYGW.jpg)